SET DATEFIRST should only accept integer arguments

jaspal007 opened the pull request:
This commit enables the check that allows only integral values to be set as a DATEFIRST values

### 1. Issues
Previously, customer could use non integral values to set the value of `DATEFIRST` 
```
1> SET DATEFIRST 6.5
2> GO
1> SELECT @@DATEFIRST
2> GO

datefirst  
-----------
          6
```
This is not an expected behavior

### 2. Changes made to fix the issue:
To fix this issue we added a check to only allow integral values to be used to set the value of `DATEFIRST`
```
1> SET DATEFIRST 6.5
2> GO
Msg 2743, Level 16, State 3, Server EC2AMAZ-GOMIGPU, Line 1
SET DATEFIRST option requires integer parameter.
```

Authored-by: Jaspal Singh ijaspals@amazon.com
Signed-off-by: Jaspal Singh ijaspals@amazon.com

### Issues Resolved

BABEL-5841

### Test Scenarios Covered ###
* **Use case based -**


* **Boundary conditions -**


* **Arbitrary inputs -**


* **Negative test cases -**


* **Minor version upgrade tests -**


* **Major version upgrade tests -**


* **Performance tests -**


* **Tooling impact -**


* **Client tests -**



### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
SET DATEFIRST should only accept integer arguments

This commit enables the check that allows only integral values to be set
as a DATEFIRST values

TASK: BABEL-5841
Signed-off-by: Jaspal Singh ijaspals@amazon.com

pranavJ23 commented on contrib/babelfishpg_tsql/src/guc.c:
```diff
@@ -607,10 +607,19 @@ assign_lock_timeout(int newval, void *extra)
+		ereport(ERROR,
+				(errcode(ERRCODE_DATATYPE_MISMATCH),
+				errmsg("SET DATEFIRST option requires integer parameter")));
+	}
```
Please add a comment explaining this logic

pranavJ23 commented on test/JDBC/expected/sys-datefirst.out:
```diff
@@ -24,3 +33,242 @@ int
+-- edge point values
+SET DATEFIRST 1.00
+GO
+SELECT @@DATEFIRST
```
Have we verified this test with tsql ?

pranavJ23 commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -478,8 +478,8 @@ datepart_internal(char* field, Timestamp timestamp, float8 df_tz, bool general_i
 		/* Adjusting the dow accourding to the datefirst guc */
-		PG_RETURN_INT32(((res) + 7 - pltsql_datefirst)%7 == 0 ?
-					7 : ((res) + 7 - pltsql_datefirst)%7);
+		PG_RETURN_INT32(((res) + 7 - (int)pltsql_datefirst)%7 == 0 ?
+					7 : ((res) + 7 - (int)pltsql_datefirst)%7);
```
Nit: please correct the indentation

pranavJ23 commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1379,7 +1379,7 @@ schema_id(PG_FUNCTION_ARGS)
 datefirst(PG_FUNCTION_ARGS)
 {
-	PG_RETURN_UINT32(pltsql_datefirst);
+	PG_RETURN_UINT32((int)pltsql_datefirst);
```
please correct indentation here as well, same as above

Commit:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into BABEL-5841

Commit:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into BABEL-5841

Commit:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into BABEL-5841

Commit:
SET DATEFIRST should only accept integer arguments

This commit enables the check that allows only integral values to be set
as a DATEFIRST values

TASK: BABEL-5841
Signed-off-by: Jaspal Singh <ijaspals@amazon.com>

Commit:
SET DATEFIRST should only accept integer arguments

This commit enables the check that allows only integral values to be set
as a DATEFIRST values

TASK: BABEL-5841
Signed-off-by: Jaspal Singh <ijaspals@amazon.com>

pranavJ23 commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -215,7 +215,7 @@ static Tuplestorestate *get_bbf_pivot_tuplestore(const char 	*sourcetext,
 extern bool is_ms_shipped(char *object_name, int type, Oid schema_id);
 
-extern int	pltsql_datefirst;
+extern int 	pltsql_datefirst;
```
Nit: unnecessary change, please correct it at all places

pranavJ23 commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -6058,6 +6058,54 @@ makeSetStatement(TSqlParser::Set_statementContext *ctx, tsqlBuilder &builder)
+				// We add a check for DATEFIRST to only accept valid input
+				if (pg_strcasecmp("DATEFIRST", val.c_str()) == 0)
+				{
+					if(ctx->set_special()->constant_LOCAL_ID())
```
instead of so many nested if's let's just throw error as : if(!ctx->set_special()->constant_LOCAL_ID()) -> error

pranavJ23 commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -6058,6 +6058,54 @@ makeSetStatement(TSqlParser::Set_statementContext *ctx, tsqlBuilder &builder)
+							// Non integral values and strings are not valid input
+							if (strchr(datefirst_val, '.') || strchr(datefirst_val, '\'') || strchr(datefirst_val, '\"') || pg_strcasecmp("NULL", datefirst_val) == 0)
+							{
+								ereport(ERROR,
```
I think let's keep this check for '.' and for the other cases let's fallback to the earlier behaviour since we were throwing error for those cases.

pranavJ23 commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -6058,6 +6058,54 @@ makeSetStatement(TSqlParser::Set_statementContext *ctx, tsqlBuilder &builder)
+								char *strtol_endptr;
+								int val = (int) strtol(datefirst_val, &strtol_endptr, 10);
+								
+								if (val < 1 || val > 7)
```
I thin kwe are already handling this case in GUC, so do we need this check here as well ?

Commit:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into BABEL-5841

Commit:
SET DATEFIRST should only accept integer arguments

This commit enables the check that allows only integral values to be set
as a DATEFIRST values

TASK: BABEL-5841
Signed-off-by: Jaspal Singh <ijaspals@amazon.com>

pranavJ23 commented on test/JDBC/expected/sys-datefirst.out:
```diff
@@ -24,3 +33,282 @@ int
+~~ERROR (Message: invalid value for parameter "babelfishpg_tsql.datefirst": "one")~~
+
+
+SET DATEFIRST '4'
+GO
+
```
we should expect an error here, right ?

Commit:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into BABEL-5841

Commit:
SET DATEFIRST should only accept integer arguments

This commit enables the check that allows only integral values to be set
as a DATEFIRST values

TASK: BABEL-5841
Signed-off-by: Jaspal Singh <ijaspals@amazon.com>

tanscorpio7 commented on contrib/babelfishpg_tsql/src/guc.c:
```diff
@@ -841,7 +841,7 @@ define_custom_variables(void)
 							gettext_noop("Sets the first day of the week to a number from 1 through 7."),
 							NULL,
-							&pltsql_datefirst,
+							&pltsql_datefirst, 
```
Not needed changes.

tanscorpio7 commented on test/JDBC/expected/sys-datefirst.out:
```diff
@@ -24,3 +33,286 @@ int
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: syntax error at or near "NULL")~~
+
```
Do we need test case where `datefirst` is in small case ?

tanscorpio7 replied:
Also please test 
```
DECLARE @i INT = 1
SET DATEFIRST @i
```

jaspal007 replied:
added the test cases

Commit:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into BABEL-5841

Commit:
SET DATEFIRST should only accept integer arguments

This commit enables the check that allows only integral values to be set
as a DATEFIRST values

TASK: BABEL-5841
Signed-off-by: Jaspal Singh <ijaspals@amazon.com>

jsudrik reviewed (approved):
Approved.
